draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+
+```
+git clone git@github.com:troessner/hello_world.git
+cd hello_world
+bundle
+```
+
+This is how you'd call `Mutant` to mutation test this:
+
+```
+bundle exec mutant --include lib/\
+                   --require hello_world\
+                   --use rspec  "Greeter*"
+```
+
+We're telling `Mutant` to:
+
+* include the "lib/" directory
+* require "hello_world.rb"
+* use RSpec as test integration
+* use the "Greeter*" namespace to find its test subjects. We'll cover the "subjects" topic later on extensively
+
+`Mutant` will then spit out part of its configuration first:
+
+```
+Mutant configuration:
+
+Matcher:         #<Mutant::Matcher::Config match_expressions: [Greeter*]>
+Integration:     Mutant::Integration::Rspec
+Expect Coverage: 100.00%
+Jobs:            4
+Includes:        ["lib/"]
+Requires:        ["hello_world"]
+Subjects:        2
+```
+
+And then `Mutant` will run your tests against the code mutations. Failing tests kill the mutants while surviving mutants show a lack of test coverage or that your code is not concise enough.
+
+This is how a surviving mutant would be indicated:
+
+```
+  evil:Greeter#say_hello:/Users/timo/dev/hello_world/lib/mutant/greeter.rb:7:80e67
+  @@ -1,6 +1,6 @@
+   def say_hello(options)
+  -  if @enabled
+  +  if true
+       "#{@phrase}#{" "}#{name}"
+     end
+   end
+```
+
+This means `Mutant` replaced the `@enabled` with an unconditional `true` and none of our tests broke.
+
+Please keep the example project and `Mutant` call from above in the back of your head - I'll refer to this example throughout the article.
+
+### Prerequisites for this article
+
+While doing research for my presentation, I wrote an introductory series of articles about the gems `Mutant` uses:
+
+* [part 1](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-1): The [Concord](https://github.com/mbj/concord) gem and the [Procto](https://github.com/snusnu/procto) gem
+* [part 2](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-2): The [IceNine](https://github.com/dkubb/ice_nine) gem
+* [part 3](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-3): The [abstract type](https://en.wikipedia.org/wiki/Abstract_type) gem and the [Adamantium](https://github.com/dkubb/adamantium) gem
+ 
+You do not need to read part 2 and 3 to fully understand everything down below but you **should** read [part 1](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-1) as I will refer to `Concord` and `Procto` quite often.
+
+`Mutant` creates its mutations by utilizing the [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) consisting of [S-expressions](https://en.wikipedia.org/wiki/S-expression) that the ruby parser generates from your source code.
```

Comment:
I think its better to say `AST represented by S-Expressions` than `consists`, as an AST does not imply you use S-expressions as input or output format, its just the most convenient format so far.

I do not want that readers falsely get into the assumption that AST implies Sexp.